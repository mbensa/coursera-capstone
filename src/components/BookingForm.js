import React, { useState, useEffect } from "react";
import Nav from "./Nav";
import Text from "./Text";
import Footer from "./Footer";
import "./bookingForm.css";

function BookingForm({
  date,
  setDate,
  name,
  setName,
  time,
  setTime,
  guests,
  setGuests,
  occasion,
  setOccasion,
  availableTimes,
  updateTimes,
  submitForm, // Receive submitForm as a prop
}) {
  const [formData, setFormData] = useState({
    date: "",
    time: "",
    guests: 1,
    occasion: "Birthday",
    name: "",
  });

  const [formValid, setFormValid] = useState(false);

  useEffect(() => {
    // Check if all form fields are valid
    const isValid = name && date && time && guests >= 1 && occasion !== "";

    // Update form validity state
    setFormValid(isValid);
  }, [date, time, guests, occasion, name]);

  const handleDateChange = (event) => {
    const newDate = event.target.value;
    setFormData((prevData) => ({
      ...prevData,
      date: newDate,
    }));
    updateTimes(newDate);
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    if (formValid) {
      submitForm({ date, time, guests, occasion, name });
    }
  };

  return (
    <>
      <Nav />
      <main className="mainContainerBooking">
        <Text type="h1">Book a table</Text>
        <div className="restaurantImg"></div>
        <form className="formContainer" onSubmit={handleSubmit}>
          <label htmlFor="res-name">
            <Text type="h5">Name and Surname</Text>
          </label>
          <input
            type="text"
            id="res-name"
            value={name}
            onChange={(e) => setName(e.target.value)}
            required
            aria-label="Name input"
          />
          <label htmlFor="res-date">
            <Text type="h5">Choose Date</Text>
          </label>
          <input
            data-testid="resDate"
            type="date"
            id="res-date"
            value={date}
            onChange={handleDateChange}
            min={new Date().toISOString().split("T")[0]}
            required
            aria-label="Date Selection"
          />
          <label htmlFor="res-time">
            <Text type="h5">Choose Time</Text>
          </label>
          <select
            id="res-time"
            value={time}
            onChange={(e) => setTime(e.target.value)}
            required
            aria-label="Time Selection"
          >
            {availableTimes.map((availableTime, index) => (
              <option key={index}>{availableTime}</option>
            ))}
          </select>
          <label htmlFor="guests">
            <Text type="h5">Number of Guests</Text>
          </label>
          <input
            type="number"
            placeholder="1"
            min="1"
            max="10"
            id="guests"
            value={guests}
            onChange={(e) => setGuests(parseInt(e.target.value))}
            required
            aria-label="Guests Selection"
          />
          <label htmlFor="occasion">
            <Text type="h5">Occassion</Text>
          </label>
          <select
            id="occasion"
            value={occasion}
            onChange={(e) => setOccasion(e.target.value)}
            required
            aria-label="Occasion Selection"
          >
            <option>Birthday</option>
            <option>Anniversary</option>
          </select>
          <input
            className="submitBtn"
            type="submit"
            value="Make Your reservation"
            disabled={!formValid}
            aria-label="Submit Reservation"
          />
        </form>
      </main>
      <Footer />
    </>
  );
}

export default BookingForm;
