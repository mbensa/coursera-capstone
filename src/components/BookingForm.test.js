import { fireEvent, render, act } from "@testing-library/react";
import BookingForm from "./BookingForm";
import fakeAPI from "./FakeAPI";
import { MemoryRouter } from "react-router-dom";
// import { initializeTimes } from "../App";
// import { updateTimes } from "../App";

const props = {
  date: new Date("2024-01-01"),
  name: "foo",
  time: "",
  guests: 1,
  occasion: "birthday",
  availableTimes: fakeAPI.fetchAPI(new Date("2024-01-01")),
};

describe("<BookingForm />", () => {
  it("should render", () => {
    const { asFragment } = render(<BookingForm {...props} />, {
      wrapper: MemoryRouter,
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders "Choose date" label', () => {
    const { getByLabelText } = render(<BookingForm {...props} />, {
      wrapper: MemoryRouter,
    });

    expect(getByLabelText("Choose Date")).toBeInTheDocument();
  });

  it("updateTimes returns the same value as provided in the state", () => {
    const mockedDate = new Date("2024-01-01T22:42:16.652Z");
    jest.spyOn(global, "Date").mockImplementation(() => mockedDate);

    const selectedDate = "2024-04-25";
    const updateTimesSpy = jest.fn();

    const { getByTestId } = render(
      <BookingForm {...props} updateTimes={updateTimesSpy} />,
      { wrapper: MemoryRouter }
    );

    act(() => {
      fireEvent.change(getByTestId("resDate"), {
        target: { value: selectedDate },
      });
    });

    expect(updateTimesSpy).toHaveBeenCalledWith(selectedDate);
  });
});
