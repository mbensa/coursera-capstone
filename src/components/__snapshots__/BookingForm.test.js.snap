// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookingForm /> should render 1`] = `
<DocumentFragment>
  <nav
    class="navbar"
  >
    <div
      class="hamburger"
    >
      <span
        class="bar"
      />
      <span
        class="bar"
      />
      <span
        class="bar"
      />
    </div>
    <svg>
      Logo.svg
    </svg>
    <ul
      class="nav-menu"
    >
      <li
        class="nav-item"
      >
        <a
          href="/"
        >
          <h5>
            Home
          </h5>
        </a>
      </li>
      <li
        class="nav-item"
      >
        <h5>
          About
        </h5>
      </li>
      <li
        class="nav-item"
      >
        <h5>
          Menu
        </h5>
      </li>
      <li
        class="nav-item"
      >
        <a
          href="/booking"
        >
          <h5>
            Reservations
          </h5>
        </a>
      </li>
      <li
        class="nav-item"
      >
        <h5>
          Order Online
        </h5>
      </li>
      <li
        class="nav-item"
      >
        <h5>
          Login
        </h5>
      </li>
    </ul>
    <svg
      class="basketLogo"
    >
      basket.svg
    </svg>
  </nav>
  <main
    class="mainContainerBooking"
  >
    <h1>
      Book a table
    </h1>
    <div
      class="restaurantImg"
    />
    <form
      class="formContainer"
    >
      <label
        for="res-name"
      >
        <h5>
          Name and Surname
        </h5>
      </label>
      <input
        aria-label="Name input"
        id="res-name"
        required=""
        type="text"
        value="foo"
      />
      <label
        for="res-date"
      >
        <h5>
          Choose Date
        </h5>
      </label>
      <input
        aria-label="Date Selection"
        data-testid="resDate"
        id="res-date"
        min="2024-04-26"
        required=""
        type="date"
        value="Mon Jan 01 2024 01:00:00 GMT+0100 (Central European Standard Time)"
      />
      <label
        for="res-time"
      >
        <h5>
          Choose Time
        </h5>
      </label>
      <select
        aria-label="Time Selection"
        id="res-time"
        required=""
      >
        <option>
          --- Select a Time ---
        </option>
        <option>
          17:00
        </option>
        <option>
          18:00
        </option>
        <option>
          19:00
        </option>
        <option>
          19:30
        </option>
        <option>
          20:00
        </option>
        <option>
          20:30
        </option>
        <option>
          21:00
        </option>
        <option>
          21:30
        </option>
        <option>
          22:30
        </option>
        <option>
          23:30
        </option>
      </select>
      <label
        for="guests"
      >
        <h5>
          Number of Guests
        </h5>
      </label>
      <input
        aria-label="Guests Selection"
        id="guests"
        max="10"
        min="1"
        placeholder="1"
        required=""
        type="number"
        value="1"
      />
      <label
        for="occasion"
      >
        <h5>
          Occassion
        </h5>
      </label>
      <select
        aria-label="Occasion Selection"
        id="occasion"
        required=""
      >
        <option>
          Birthday
        </option>
        <option>
          Anniversary
        </option>
      </select>
      <input
        aria-label="Submit Reservation"
        class="submitBtn"
        disabled=""
        type="submit"
        value="Make Your reservation"
      />
    </form>
  </main>
  <footer
    class="footerContainer"
  >
    <svg>
      Logo.svg
    </svg>
    <div
      class="socialContainer"
    >
      <svg
        class="facebook social"
      >
        facebook.svg
      </svg>
      <svg
        class="youtube social"
      >
        youtube.svg
      </svg>
      <svg
        class="twitter social"
      >
        twitter.svg
      </svg>
      <svg
        class="instagram social"
      >
        instagram.svg
      </svg>
    </div>
  </footer>
</DocumentFragment>
`;
